crates/e2e/src/client.rs

```diff
@@ -307,12 +310,20 @@ where
             Error::CallDryRun(_) => f.write_str("CallDryRun"),
             Error::CallExtrinsic(_) => f.write_str("CallExtrinsic"),
             Error::Balance(msg) => write!(f, "Balance: {msg}"),
+            Error::Decoding(err) => write!(f, "Decoding: {err}"),
         }
     }
 }
 
 /// A contract was successfully instantiated.
-#[derive(Debug, scale::Decode, scale::Encode)]
+#[derive(
+    Debug,
+    scale::Decode,
+    scale::Encode,
+    scale_decode::DecodeAsType,
+    scale_encode::EncodeAsType,
+)]
+#[decode_as_type(trait_bounds = "")]
```

Comment:
Why do you need this if trait bounds are empty?

Reply:
because automatically compiler will require bounds for `E` (in general, for every generic parameter)

Reply:
Perhaps to avoid having this attribute everywhere we could have a blanket impl of `DecodeAsType` for the `Environment` trait?

Reply:
we could, but then I would have to provide something reasonable here:
```rust
impl DecodeAsType for E {
    fn decode_as_type(input: &mut &[u8], type_id: u32, types: &PortableRegistry) -> Result<Self, Error> {
        todo!()
    }
}

impl EncodeAsType for E {
    fn encode_as_type_to(&self, type_id: u32, types: &PortableRegistry, out: &mut Vec<u8>) -> Result<(), scale_encode::Error> {
        todo!()
    }
}
```

@jsdw what makes sense here?

Reply:
Assuming `Environment` is a type that can't be constructed (eg `enum Environment {}`) I think some like that would be fine (though I'd `panic!("Shouldn't call x on Environment")` on the `DecodeAsType` and maybe `unreachable!()` on the `EncodeAsType` one even so).

If it's at all possible to construct an `Environment` then personally I'd go for the trait bounds everywhere.

(Side note: I'm noting that `Encode` and `Decode` don't require such trait bounds; I wonder how that is possible because maybe we can be smarter with `EncodeAsType` and `DecodeAsType` in this regard also)

Reply:
hmm... it's a trait and it feels kinda weird to have these panics in the blanket implementation :| I don't have strong opinion; @ascjones (?)

Reply:
Not a good idea. Just leave as is with the `#[decode_as_type(trait_bounds = "")]` for now